arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,3181 @@
+        }
+
+      /* Clear txb status change interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_TXBHCI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Check if any of the stuck one belongs to RX buffer data overrun */
+
+  if (isr & MPFS_CANFD_INT_STAT_DOI)
+    {
+      struct canfd_frame *cf = (struct canfd_frame *)priv->rxdesc;
+
+      caninfo("rx fifo overflow\n");
+
+      cf->can_id = CAN_ERR_CRTL;
+      cf->data[1] = CAN_ERR_CRTL_RX_OVERFLOW;
+
+      /* Copy the buffer pointer to priv->dev..  Set amount of data
+       * in priv->dev.d_len
+       */
+
+      priv->dev.d_len = sizeof(struct canfd_frame);
+      priv->dev.d_buf = (uint8_t *)cf;
+
+      /* Send to socket interface */
+
+      NETDEV_RXPACKETS(&priv->dev);
+      can_input(&priv->dev);
+
+      /* Point the packet buffer back to the next Tx buffer that will be
+       * used during the next write.
+       */
+
+      priv->dev.d_buf = (uint8_t *)priv->txdesc;
+
+      /* Clear Data Overrun interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_DOI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Clear and reset all interrupt. */
+
+  caninfo("Reset all interrupts...\n");
+  imask = 0xffffffff;
+  putreg32(imask, priv->base + MPFS_CANFD_INT_ENA_CLR_OFFSET);
+  putreg32(imask, priv->base + MPFS_CANFD_INT_ENA_SET_OFFSET);
+  return OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_get_tx_status
+ *
+ * Description:
+ *  Get status of txt buffer
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private FPGA CANFD driver state structure
+ *  buf   - txt buffer index to get status of (0-based)
+ *
+ * Returned Value:
+ *  Status of txt buffer
+ *
+ * Assumptions:
+ *  None
+ *
+ ****************************************************************************/
+
+static enum mpfs_can_txtb_status
+  mpfs_can_get_tx_status(FAR struct mpfs_driver_s *priv,
+                         uint8_t buf)
+{
+  uint32_t tx_status = getreg32(priv->base + MPFS_CANFD_TX_STATUS_OFFSET);
+  enum mpfs_can_txtb_status status = (tx_status >> (buf * 4)) & 0xf;
+
+  return status;
+}
+
+/****************************************************************************
+ * Name: mpfs_can_is_txt_buf_writable
+ *
+ * Description:
+ *  Check if frame can be inserted to TXT Buffer
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private FPGA CANFD driver state structure
+ *  buf   - txt buffer index to get status of (0-based)
+ *
+ * Returned Value:
+ *  True  - Frame can be inserted to txt buffer
+ *  False - If attempted, frame will not be inserted to txt buffer
+ *
+ * Assumptions:
+ *  None
+ *
+ ****************************************************************************/
+
+static bool mpfs_can_is_txt_buf_writable(FAR struct mpfs_driver_s *priv,
+                                         uint8_t buf)
+{
+  enum mpfs_can_txtb_status buf_status;
+
+  buf_status = mpfs_can_get_tx_status(priv, buf);
```

Comment:
Very difficult to understand what happens here, so do you have like n-buffers for TX, and you need to see per index if a buffer has been handled or no ? 

Why not just use a freelist for free buffers and a sendlist for buffers that need sending. Another option is a FIFO. Maybe this is just some detail with the CAN IP I don't understand, but like I said, this is not immediately obvious why the per-buffer states are needed ?

Reply:
Is this similar to ethernet mac dma; i.e. the HW scans for the buffers and sends them out, after which it writes the buffer status. And SW just checks if the buffer has already been sent out (i.e. it is free to write new data into)?